third_party/boringssl/blake2b512.patch

```diff
@@ -0,0 +1,144 @@
++  data += todo;
++  len -= todo;
++
++  if (!len) {
++    return;
++  }
++
++  // More input remains therefore we must have filled |b2b->block|.
++  assert(b2b->block_used == BLAKE2B_CBLOCK);
++  blake2b_transform(b2b, b2b->block, BLAKE2B_CBLOCK,
++                    /*is_final_block=*/0);
++  b2b->block_used = 0;
++
++  while (len > BLAKE2B_CBLOCK) {
++    blake2b_transform(b2b, data, BLAKE2B_CBLOCK, /*is_final_block=*/0);
++    data += BLAKE2B_CBLOCK;
++    len -= BLAKE2B_CBLOCK;
++  }
++
++  OPENSSL_memcpy(b2b->block, data, len);
++  b2b->block_used = len;
++}
++
+ void BLAKE2B256_Final(uint8_t out[BLAKE2B256_DIGEST_LENGTH], BLAKE2B_CTX *b2b) {
+   OPENSSL_memset(&b2b->block[b2b->block_used], 0,
+                  sizeof(b2b->block) - b2b->block_used);
+@@ -160,6 +213,15 @@ void BLAKE2B256_Final(uint8_t out[BLAKE2B256_DIGEST_LENGTH], BLAKE2B_CTX *b2b) {
+   memcpy(out, b2b->h, BLAKE2B256_DIGEST_LENGTH);
+ }
+
++void BLAKE2B512_Final(uint8_t out[BLAKE2B512_DIGEST_LENGTH], BLAKE2B_CTX *b2b) {
++  OPENSSL_memset(&b2b->block[b2b->block_used], 0,
++                 sizeof(b2b->block) - b2b->block_used);
++  blake2b_transform(b2b, b2b->block, b2b->block_used,
++                    /*is_final_block=*/1);
++  static_assert(BLAKE2B512_DIGEST_LENGTH <= sizeof(b2b->h), "");
++  memcpy(out, b2b->h, BLAKE2B512_DIGEST_LENGTH);
++}
++
+ void BLAKE2B256(const uint8_t *data, size_t len,
+                 uint8_t out[BLAKE2B256_DIGEST_LENGTH]) {
+   BLAKE2B_CTX ctx;
+@@ -167,3 +229,11 @@ void BLAKE2B256(const uint8_t *data, size_t len,
+   BLAKE2B256_Update(&ctx, data, len);
+   BLAKE2B256_Final(out, &ctx);
+ }
++
++void BLAKE2B512(const uint8_t *data, size_t len,
++                uint8_t out[BLAKE2B512_DIGEST_LENGTH]) {
++  BLAKE2B_CTX ctx;
++  BLAKE2B512_Init(&ctx, NULL);
++  BLAKE2B512_Update(&ctx, data, len);
++  BLAKE2B512_Final(out, &ctx);
++}
+diff --git a/src/include/openssl/blake2.h b/src/include/openssl/blake2.h
+index 03e3a465c..8bd409d77 100644
+--- a/src/include/openssl/blake2.h
++++ b/src/include/openssl/blake2.h
+@@ -21,9 +21,10 @@
+ extern "C" {
+ #endif
+
+-
+ #define BLAKE2B256_DIGEST_LENGTH (256 / 8)
++#define BLAKE2B512_DIGEST_LENGTH (512 / 8)
+ #define BLAKE2B_CBLOCK 128
++#define BLAKE2B_PERSONAL_LENGTH 16
+
+ struct blake2b_state_st {
+   uint64_t h[8];
+@@ -51,6 +52,28 @@ OPENSSL_EXPORT void BLAKE2B256_Final(uint8_t out[BLAKE2B256_DIGEST_LENGTH],
+ OPENSSL_EXPORT void BLAKE2B256(const uint8_t *data, size_t len,
+                                uint8_t out[BLAKE2B256_DIGEST_LENGTH]);
+
++// BLAKE2B512_Init initialises |b2b| to perform a BLAKE2b-512 hash. There are no
++// pointers inside |b2b| thus release of |b2b| is purely managed by the caller.
++// |personal|, if non-NULL, must point to |BLAKE2B_PERSONAL_LENGTH| bytes of
++// personalization data. |personal| may be NULL in which case no personalization
++// is used.
++OPENSSL_EXPORT void BLAKE2B512_Init(BLAKE2B_CTX *b2b, const char personal[BLAKE2B_PERSONAL_LENGTH]);
```

Comment:
I think it's good to have both interfaces considering that `BLAKE2B512()` just initializes without personal.

```c++
OPENSSL_EXPORT void BLAKE2B512_Init(BLAKE2B_CTX *b2b);
OPENSSL_EXPORT void BLAKE2B512_InitWithPersonal(BLAKE2B_CTX *b2b,
                                    const char personal[BLAKE2B_PERSONAL_LENGTH]);
```